examples/generative/stylegan.py

```diff
@@ -0,0 +1,736 @@
+
+
+"""
+### Utility function to display images after each epoch
+"""
+
+
+def plot_images(images, log2_res, fname=""):
+    scales = {2: 0.5, 3: 1, 4: 2, 5: 3, 6: 4, 7: 5, 8: 6, 9: 7, 10: 8}
+    scale = 0.5 * scales[log2_res]
+
+    grid_col = min(images.shape[0], int(16 // scale))
+    grid_row = 1
+
+    f, axarr = plt.subplots(
+        grid_row, grid_col, figsize=(grid_col * scale, grid_row * scale)
+    )
+
+    for row in range(grid_row):
+        ax = axarr if grid_row == 1 else axarr[row]
+        for col in range(grid_col):
+            ax[col].imshow(images[row * grid_col + col])
+            ax[col].axis("off")
+    plt.show()
+    if fname:
+        f.savefig(fname)
+
+
+"""
+## Custom Layers
+The following are building blocks that will be used to construct the generators and
+discriminators.
+"""
+
+
+def fade_in(alpha, a, b):
+
+    return alpha * a + (1.0 - alpha) * b
+
+
+def wasserstein_loss(y_true, y_pred):
+
+    return -tf.reduce_mean(y_true * y_pred)
+
+
+class PixelNorm(Layer):
+    def __init__(self, epsilon=1e-8):
+        super(PixelNorm, self).__init__()
+        self.epsilon = epsilon
+
+    def call(self, input_tensor):
+        return input_tensor / tf.math.sqrt(
+            tf.reduce_mean(input_tensor ** 2, axis=-1, keepdims=True) + self.epsilon
+        )
+
+
+class MinibatchStd(Layer):
+    def __init__(self, group_size=4, epsilon=1e-8):
+
+        super(MinibatchStd, self).__init__()
+        self.epsilon = epsilon
+        self.group_size = group_size
+
+    def call(self, input_tensor):
+
+        n, h, w, c = input_tensor.shape
+        x = tf.reshape(input_tensor, [self.group_size, -1, h, w, c])
+        group_mean, group_var = tf.nn.moments(x, axes=(0), keepdims=False)
+        group_std = tf.sqrt(group_var + self.epsilon)
+        avg_std = tf.reduce_mean(group_std, axis=[1, 2, 3], keepdims=True)
+        x = tf.tile(avg_std, [self.group_size, h, w, 1])
+
+        return tf.concat([input_tensor, x], axis=-1)
+
+
+class ConvBlock(Layer):
+    def __init__(self, out_channels, kernel=3, gain=2, **kwargs):
+        super(ConvBlock, self).__init__(kwargs)
+        self.kernel = kernel
+        self.out_channels = out_channels
+        self.gain = gain
+        self.pad = kernel != 1
+
+    def build(self, input_shape):
+        self.in_channels = input_shape[-1]
+
+        initializer = tf.keras.initializers.RandomNormal(mean=0.0, stddev=1.0)
+        self.w = self.add_weight(
+            shape=[self.kernel, self.kernel, self.in_channels, self.out_channels],
+            initializer=initializer,
+            trainable=True,
+            name="kernel",
+        )
+
+        self.b = self.add_weight(
+            shape=(self.out_channels,), initializer="zeros", trainable=True, name="bias"
+        )
+
+        fan_in = self.kernel * self.kernel * self.in_channels
+        self.scale = tf.sqrt(self.gain / fan_in)
+
+    def call(self, inputs):
+        if self.pad:
```

Comment:
How does this differ from `SAME` mode?

Reply:
`SAME` always pad it with zeros? The `self.pad` is only False for 1x1 conv so reflective padding is used by default.